lib/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEExtendedTaintAnalysis.cpp

```diff
@@ -0,0 +1,841 @@
+}
+
+void IDEExtendedTaintAnalysis::identity(std::set<d_t> &Ret, const d_t &Source,
+                                        const llvm::Instruction *CurrInst,
+                                        bool AddGlobals) {
+
+  if (AddGlobals && llvm::isa<llvm::GlobalValue>(Source->base())) {
+    Ret.insert(Source);
+  } else if (const auto *Inst =
+                 llvm::dyn_cast<llvm::Instruction>(Source->base());
+             Inst && Inst->getFunction() == CurrInst->getFunction()) {
+    Ret.insert(Source);
+  } else if (const auto *Arg = llvm::dyn_cast<llvm::Argument>(Source->base());
+             Arg && Arg->getParent() == CurrInst->getFunction()) {
+    Ret.insert(Source);
+  }
+}
+
+auto IDEExtendedTaintAnalysis::identity(const d_t &Source,
+                                        const llvm::Instruction *CurrInst,
+                                        bool AddGlobals) -> std::set<d_t> {
+  std::set<d_t> Ret;
+  identity(Ret, Source, CurrInst, AddGlobals);
+  return Ret;
+}
+
+auto IDEExtendedTaintAnalysis::transferFlowFact(d_t Source, d_t From,
+                                                const llvm::Value *To) -> d_t {
+
+  return FactFactory.withTransferTo(Source, From, To);
+}
+
+const llvm::Instruction *
+IDEExtendedTaintAnalysis::getApproxLoadFrom(const llvm::Instruction *V) const {
+
+  if (llvm::isa<llvm::LoadInst>(V) || llvm::isa<llvm::CallBase>(V)) {
+    return V;
+  }
+
+  if (const auto *It = V->op_begin();
+      It != V->op_end() && llvm::isa<llvm::Instruction>(*It)) {
+    return getApproxLoadFrom(llvm::cast<llvm::Instruction>(*It));
+  }
+
+  return V;
+}
+
+const llvm::Instruction *
+IDEExtendedTaintAnalysis::getApproxLoadFrom(const llvm::Value *V) const {
+  if (V->getType()->isPointerTy()) {
+    return nullptr;
+  }
+
+  if (const auto *Inst = llvm::dyn_cast<llvm::Instruction>(V)) {
+    return getApproxLoadFrom(Inst);
+  }
+
+  return nullptr;
+}
+
+void IDEExtendedTaintAnalysis::doPostProcessing(
+    const SolverResults<n_t, d_t, l_t> &SR) {
+  PostProcessed = true;
+  llvm::SmallVector<const llvm::Instruction *> RemInst;
+  for (auto &[Inst, PotentialLeaks] : Leaks) {
+    llvm::SmallVector<const llvm::Value *, 2> Rem;
+    std::cerr << "At " << llvmIRToString(Inst) << ":" << std::endl;
+
+    auto Results = SR.resultsAt(Inst);
+
+    for (const auto *L : PotentialLeaks) {
+      auto Found = Results.find(makeFlowFact(L));
+      if (Found == Results.end()) {
+        // The sanitizer has been killed, so we must assume the fact as tainted
+        std::cerr << "No results for " << makeFlowFact(L) << std::endl;
+        continue;
+      }
+
+      auto Sani = // SR.resultAt(Inst, makeFlowFact(L));
+          Found->second;
+      const auto *Load = getApproxLoadFrom(L);
+
+      switch (Sani.getKind()) {
+      // case EdgeDomain::Bot:
+      //   rem.push_back(L);
+      //   std::cerr << "Sanitize " << llvmIRToShortString(L) << " with Bottom "
+      //             << std::endl;
+      //   break;
```

Comment:
Is this case no longer needed?

Reply:
In the current design, Bottom is no proper sanitizer; so this can be removed